dwave/optimization/generators.py

```diff
@@ -407,6 +409,243 @@ class as the decision variable being optimized, with permutations of its
+    time_close_ext = np.zeros(num_customers + 1)
+    time_close_ext[:num_customers] = time_window_close[1:]
+    time_close_ext[-1] = 1e6
+    t_close = model.constant(time_close_ext)
+
+    service_time_ext = np.zeros(num_customers + 1)
+    service_time_ext[:num_customers] = service_time[1:]
+    t_service = model.constant(service_time_ext)
+
+    demand = model.constant(customer_demand)
+    capacity = model.constant(vehicle_capacity)
+    t_depo_close = model.constant(time_window_close[0])
+
+    one = model.constant(1)
+
+    # Add the decision variable
+    routes_decision, routes = model.disjoint_lists(
+        primary_set_size=num_customers,
+        num_disjoint_lists=number_of_vehicles)
+
+    # Capacity constraint
+    capacity_constraints = [(demand[routes[vehicle_idx]].sum() <= capacity)
+                            for vehicle_idx in range(number_of_vehicles)
+                            ]
+    for c in capacity_constraints:
+        model.add_constraint(c)
+
+    rh = np.arange(num_customers + 1).astype(int)
+    range_helper = model.constant(rh)
+
+    # this is number of clients each vehicle visits
+    num_clients_in_route = {}
+    for i in range(number_of_vehicles):
+        num_clients_in_route[f'route{i}'] = routes[i].size()
+
+    # Constrain the number of locations per route
+    max_loc_per_route_constant = model.constant(3 * int(num_customers / number_of_vehicles))
+    max_loc_per_route_constraints = [(num_clients_in_route[f'route{v}'] <= max_loc_per_route_constant)
+                                     for v in range(number_of_vehicles)]
+    for mlpr in max_loc_per_route_constraints:
+        model.add_constraint(mlpr)
+
+    # here we keep track of the time a vehicle arrives to each client
+    # store them in the class to check the solution
+    time_leaving = []
+    serving_time = []
+    times_back = []
+    time_windows_constraints = []
+
+    for vehicle_idx in range(number_of_vehicles):
+        this_t_leaving = []
+        this_t_windows_c = []
+        this_t_serving_time = []
+
+        for client_idx in range(max_loc_per_route_constant):
+
+            # the condition here allows us to choose whether to choose a used index or to fall back to the last one
+            # that points to a zero in the extended array
+            condition = (num_clients_in_route[f'route{vehicle_idx}'] >= range_helper[client_idx] + one)
+
+            if client_idx == 0:
+                # index of the first customer visited, if length of the route is zero,
+                # points to the last (0) element
+                idx = where(condition, routes[vehicle_idx][:1].sum(), range_helper[-1])
+                this_t_serving_time.append(maximum(t_depo[idx], t_open[idx]))
+                this_t_leaving.append(this_t_serving_time[-1] + t_service[idx])
+
+            else:
+                # index of the previous customer visited
+                previous_idx = (condition * routes[vehicle_idx][client_idx - 1:client_idx].sum()
```

Comment:
Why aren't we using a `where` here? I see that replacing this statement with
```python
                previous_idx = where(condition, routes[vehicle_idx][client_idx - 1:client_idx].sum(), range_helper[-1])
```
leads to test failures. But shouldn't it be the same? I feel like I am missing something obvious.

Reply:
Oh, duh, the test that's failing is checking the number of nodes. So that's expected.

Reply:
Thank you, I missed this line when adding where nodes. Will fix this.